### PR TITLE
[`RenderEditable`] report real height when `maxLines == 1`. 

### LINKED ISSUE #112123: SkParagraph doesn't account for `applyHeightToFirstAscent` `applyHeightToLastDescent` on empty paragraphs with strut enabled
## Steps to Reproduce

```dart
const style = TextStyle(height: 11, fontSize: 7);
const simple = TextSpan(text: 'x', style: style);
const emptyString = TextSpan(text: '', style: style);
const emptyParagraph = TextSpan(style: style);

final painter = TextPainter(
  textDirection: TextDirection.ltr,
  strutStyle: StrutStyle.fromTextStyle(style),
  textHeightBehavior: const TextHeightBehavior(applyHeightToFirstAscent: false, applyHeightToLastDescent: false),
);

for (final span in [simple, emptyString, emptyParagraph]) {
  painter.text = span;
  painter.layout();
  final realHeight = painter.height;
  final preferredHeight = painter.preferredLineHeight;
  print('${span.text} > real height: $realHeight, preferred height: $preferredHeight');
}
```
"real height" should always be the same as "preferred height" (8 px) in all 3 cases. However right now for both "emptyString" and "emptyParagraph" the reported "real height" is 77.0.

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Another one extracted from #110966. 

When `maxLines == 1` `RenderEditable` can't be scrolled vertically to reveal clipped contents so it must report its real height (instead of the estimated height `preferredLineHeight`). I'll try to fix https://github.com/flutter/flutter/issues/112123 after this and removes the hack in a later pull request.

~Running TGP.~ TGP passed with no scuba changes.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 1]
```diff
@@ -1160,7 +1160,7 @@ class TextPainter {
   /// Valid only after [layout] has been called.
   List<ui.LineMetrics> computeLineMetrics() {
     assert(_debugAssertTextLayoutIsValid);
-    return  _lineMetricsCache ??= _paragraph!.computeLineMetrics();
+    return _lineMetricsCache ??= _paragraph!.computeLineMetrics();
```
Comment (justinmc) [inline packages/flutter/lib/src/painting/text_painter.dart:1163]: 👍

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -1790,42 +1797,64 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
...
+  int? _cachedLineBreakCount;
+  int _countHardLineBreaks(String text) {
+    final int? cachedValue = _cachedLineBreakCount;
+    if (cachedValue != null) {
+      return cachedValue;
+    }
+    int count = 0;
+    for (int index = 0; index < text.length; index += 1) {
+      // TODO(LongCatIsLooong): account for other line terminators, or get it
+      // from ui.Paragraph.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:1809]: Did you mean to handle this in this PR?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 3]
```diff
@@ -1790,42 +1797,64 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
...
+      count += text.codeUnitAt(index) == 0x0A ? 1 : 0;
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:1810]: Nit: Maybe make 0x0A a constant for clarity about what it is, depending on your TODO above.

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 4]
```diff
@@ -31,43 +31,32 @@ abstract class RenderShiftedBox extends RenderBox with RenderObjectWithChildMixi
...
+    return child?.getMinIntrinsicHeight(width) ?? 0.0;
   }
 
   @override
   double computeMaxIntrinsicHeight(double width) {
-    if (child != null) {
-      return child!.getMaxIntrinsicHeight(width);
-    }
-    return 0.0;
+    return child?.getMaxIntrinsicHeight(width) ?? 0.0;
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/shifted_box.dart:49]: 👏

**Review by justinmc — APPROVED** [review]
LGTM 👍 but take a look at your TODO below that you might want to still handle in this PR.

Thanks for fixing this and for splitting it out of your other PR.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1809]: Added other line breakers in. But still it would be better to let skparagraph handle that.

**Pull request merged**